Use store based barrier in init_process_group.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49577 Add benchmakr for torch.distributed.pipeline.sync.Pipe
* #48638 Improve documentation for pipeline parallelism.
* #48630 Cleanup APIs for pipeline parallelism.
* #48470 Test pipeline parallelism works with DDP.
* **#49419 Use store based barrier in init_process_group.**

As described in https://github.com/pytorch/pytorch/issues/48110, the
newly introduced `barrier()` in `init_process_group` messes up NCCL
communicator state since it uses a bunch of default devices to perform an
allreduce which simulates a barrier(). As a ressult, subsequent NCCL operations
might not behave as expected.

Differential Revision: [D25566550](https://our.internmc.facebook.com/intern/diff/D25566550/)

pritamdamania committed:
Use store based barrier in init_process_group.

As described in https://github.com/pytorch/pytorch/issues/48110, the
newly introduced `barrier()` in `init_process_group` messes up NCCL
communicator state since it uses a bunch of default devices to perform an
allreduce which simulates a barrier(). As a ressult, subsequent NCCL operations
might not behave as expected.

Differential Revision: [D25566550](https://our.internmc.facebook.com/intern/diff/D25566550/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -172,6 +172,15 @@ class GroupMember(object):
+def _store_based_barrier(rank, store):
+    global _group_count
+    store_key = "{}:{}".format(STORE_BASED_BARRIER_PREFIX, _group_count)
+    if rank == 0:
+        store.set(store_key, "1")
```
This probably won't work since rank 0 doesn't wait for all other ranks, will fix this.

pritamdamania committed:
Update on "Use store based barrier in init_process_group."


As described in https://github.com/pytorch/pytorch/issues/48110, the
newly introduced `barrier()` in `init_process_group` messes up NCCL
communicator state since it uses a bunch of default devices to perform an
allreduce which simulates a barrier(). As a ressult, subsequent NCCL operations
might not behave as expected.

Differential Revision: [D25566550](https://our.internmc.facebook.com/intern/diff/D25566550/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2443,6 +2466,12 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None):
     # process groups including global variables are updated correctly on all
     # ranks.
-    barrier()
+    if backend == Backend.MPI:
```
there might be other c10d extensions besides MPI. Does it make sense if we only do `_store_based_barrier` for NCCL backend?

pritamdamania87 replied:
This follows the same logic in other parts of the code where MPI assumes there is no store and all other backends have to initialize a store.

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -172,6 +173,22 @@ class GroupMember(object):
+STORE_BASED_BARRIER_PREFIX = "store_based_barrier_key"
+
+def _store_based_barrier(rank, store, timeout):
+    global _group_count
```
Looks like we are only reading from `_group_count`. If that's the case, shall we drop this line?

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -172,6 +173,22 @@ class GroupMember(object):
+    world_size = get_world_size()
+    worker_count = int(store.get(store_key))
+    start = time.time()
+    while worker_count != world_size:
```
this means `_store_based_barrier` can only be called once for each pg? Shall we throw proper error if somehow `worker_count` becomes larger than `world_size`?

pritamdamania87 replied:
`_store_based_barrier` is a private method for now and only intended for use with init_process_group/new_group which is only done once per process group.

mrshenli commented on test/distributed/test_c10d.py:
```diff
@@ -4528,6 +4528,77 @@ def test_broadcast_coalesced_gloo_cpu(self):
+        c10d.all_reduce(t)
+
+        pg = c10d.new_group([1])
+        c10d.all_reduce(t)
```
nit: do we also want to assert on the results in this test?

mrshenli reviewed (approved):
LGTM! Added some minor comments.

pritamdamania committed:
Update on "Use store based barrier in init_process_group."


As described in https://github.com/pytorch/pytorch/issues/48110, the
newly introduced `barrier()` in `init_process_group` messes up NCCL
communicator state since it uses a bunch of default devices to perform an
allreduce which simulates a barrier(). As a ressult, subsequent NCCL operations
might not behave as expected.

Differential Revision: [D25566550](https://our.internmc.facebook.com/intern/diff/D25566550/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Use store based barrier in init_process_group."


As described in https://github.com/pytorch/pytorch/issues/48110, the
newly introduced `barrier()` in `init_process_group` messes up NCCL
communicator state since it uses a bunch of default devices to perform an
allreduce which simulates a barrier(). As a ressult, subsequent NCCL operations
might not behave as expected.

Differential Revision: [D25566550](https://our.internmc.facebook.com/intern/diff/D25566550/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Use store based barrier in init_process_group."


As described in https://github.com/pytorch/pytorch/issues/48110, the
newly introduced `barrier()` in `init_process_group` messes up NCCL
communicator state since it uses a bunch of default devices to perform an
allreduce which simulates a barrier(). As a ressult, subsequent NCCL operations
might not behave as expected.

Differential Revision: [D25566550](https://our.internmc.facebook.com/intern/diff/D25566550/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Use store based barrier in init_process_group."


As described in https://github.com/pytorch/pytorch/issues/48110, the
newly introduced `barrier()` in `init_process_group` messes up NCCL
communicator state since it uses a bunch of default devices to perform an
allreduce which simulates a barrier(). As a ressult, subsequent NCCL operations
might not behave as expected.

Differential Revision: [D25566550](https://our.internmc.facebook.com/intern/diff/D25566550/)

[ghstack-poisoned]
